include/unifex/indexed_for.hpp

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/config.hpp>
+#include <unifex/receiver_concepts.hpp>
+#include <unifex/sender_concepts.hpp>
+#include <unifex/type_traits.hpp>
+#include <unifex/blocking.hpp>
+#include <unifex/get_stop_token.hpp>
+#include <unifex/async_trace.hpp>
+
+#include <functional>
+#include <type_traits>
+
+namespace execution {
+class sequenced_policy;
+class parallel_policy;
+}
+
+namespace unifex {
+
+template <typename Predecessor, typename Policy, typename Range, typename Func>
+struct indexed_for_sender {
+  UNIFEX_NO_UNIQUE_ADDRESS Predecessor pred_;
+  UNIFEX_NO_UNIQUE_ADDRESS Policy policy_;
+  UNIFEX_NO_UNIQUE_ADDRESS Range range_;
+  UNIFEX_NO_UNIQUE_ADDRESS Func func_;
+
+  template <template <typename...> class Tuple>
+  struct indexed_for_result {
+   public:
+    template <typename... Args>
+    using apply = Tuple<Args...>;
+  };
+
+  template <typename... Args>
+  using is_overload_noexcept = std::bool_constant<noexcept(
+      std::invoke(std::declval<Func>(), std::declval<Args>()...))>;
+
+  template <template <typename...> class Variant>
+  struct calculate_errors {
+   public:
+    template <typename... Errors>
+    using apply = std::conditional_t<
```

Comment:
I think in general we need to always add std::exception_ptr to the list of error_types here since we don't yet know whether or not the output-receiver will have a noexcept `set_value()` or not.

This is a side-effect of allowing `set_value()` to be potentially-throwing.

Once I make the change to allow also putting the `value_types` and `error_types` onto the operation-state, then we can more accurately deduce whether or not something is non-throwing or not.
The operation_state::error_types might be able to report a reduced subset of the error types reported by the sender::error_types member.

So, for now, we can probably simplify this to remove the conditional_t and just always append exception_ptr.

Reply:
Makes sense. Copy/paste from transform though so this is a wider change that should be made later.